src/Npgsql/NpgsqlConnector.cs

```diff
@@ -1549,19 +1551,29 @@ void DoCancelRequest(int backendProcessId, int backendSecretKey)
         }
 
         /// <summary>
-        /// Resets cancellation-related state to prepare for a new cancellable operation.
+        /// Starts an operation that can be cancelled by the user or time out.
         /// </summary>
         /// <param name="userCancellationToken">
-        /// The cancellation token provided by the user for the operation. This is stored on the connector, and will be referenced
-        /// by the <see cref="OperationCanceledException"/> if one is thrown.
+        /// The cancellation token provided by the user. Callbacks will be registered on this token for executing the cancellation,
+        /// and the token will be included in any thrown <see cref="OperationCanceledException"/>.
+        /// </param>
+        /// <param name="attemptPgCancellation">
+        /// If <see langword="true" />, PostgreSQL cancellation will be attempted when the user requests cancellation or a timeout
+        /// occurs, followed by a client-side socket cancellation once <see cref="NpgsqlConnectionStringBuilder.CancellationTimeout"/> has
+        /// elapsed. If <see langword="false" />, PostgreSQL cancellation will be skipped and client-socket cancellation will occur
+        /// immediately.
         /// </param>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        internal void ResetCancellation(CancellationToken userCancellationToken)
+        internal CancellationTokenRegistration StartCancellableOperation(
```

Comment:
I wonder, what if we just check for for the default cancellation token, instead of passing `startCancellableOperation` flag?

Reply:
Or just check `CanBeCancelled`.

Reply:
Things look a bit different in my latest branch work, wait a little bit :)

BTW calling Register on a non-cancellable token is a very fast operations - basically the same as checking CanBeCanceled. In that case the default CancellationTokenRegistration is returned, which is also very fast to Dispose (one check and then do nothing). So in my current work I'm not bothering any more with checking tokens and having nullable registrations...